File: 32.js
function nextGreaterElements(nums) {
  const n = nums.length;
  const res = Array(n).fill(-1);
  const stack = []; // stores indices

  for (let i = 0; i < 2 * n; i++) {
    while (stack.length && nums[stack[stack.length - 1]] < nums[i % n]) {
      const idx = stack.pop();
      res[idx] = nums[i % n];
    }
    if (i < n) stack.push(i);
  }
  return res;
}

// Example
console.log(nextGreaterElements([1, 2, 1])); // [2, -1, 2]

File: 32.py
def nextGreaterElements(nums):
    n = len(nums)
    res = [-1] * n
    stack = []  # store indices

    # traverse twice to simulate circular array
    for i in range(2 * n):
        while stack and nums[stack[-1]] < nums[i % n]:
            idx = stack.pop()
            res[idx] = nums[i % n]
        if i < n:
            stack.append(i)
    return res


# Example
nums = [1, 2, 1]
print(nextGreaterElements(nums))  # Output: [2, -1, 2]
